### ISSUE TITLE
Feature Request: Customizable keybindings

### ISSUE THREAD

**Issue by leosolid** [excalidraw/excalidraw#3852]
I really want to move all binds to the left side of the keyboard, so I can access them all without moving my left hand. I've been looking for a whiteboard app that can do this and I couldn't find anything, so I might have to fork this and add it myself.

**Comment by ad1992** [comment]
@leosolid we have a similar feature request in https://github.com/excalidraw/excalidraw/issues/3706, we will be adding this capability soon.

**Comment by dwelle** [comment]
I think @leosolid is primarily talking about ability to change shortcuts in-app? If so, then this will require a rewrite of how we're handling shortcut — which is already planned.

We'll investigate if customizing shortcuts can be added at the same time.

**Comment by ad1992** [comment]
Yeah, I was thinking of adding the capability to allow the host to have the capability to map the shortcuts, this will solve the crashing issues as well as give them the freedom to customize it the way they need.

**Comment by leosolid** [comment]
@ad1992 @dwelle Awesome, I'll leave to you guys then. If there is a PR I can follow, please do share it.

**Comment by Fierce-Ninja** [comment]
+1 on this.  Already upvoted it.  Would love to see this 🙂

**Comment by iancustoica** [comment]
yep, this be a very much needed feature me thinks.

**Comment by IEatCodeDaily** [comment]
+1 very very important QoL.
or at least please add a keyboard shortcut for "Insert ANY file". I really need it as I import a lot of .SVG whenever I sketch something in excalidraw.

**Comment by louremipsum** [comment]
**I wanted to and am working on this issue. I just wanted to confirm with the team if the feature is already being implemented or if I am good to go. I wanted to ask before as I saw comments in various related issue that it is going to be worked on by members of Excalidraw but those comments were 2-3 yrs old**

## Plan of action(very initial)

I was thinking of first adding a new sub-menu named "Keyboard shortcuts" and inside that, a person can see the default keyboard shortcuts as well as edit them. 
To edit an existing keyboard shortcut, they would have to click the particular shortcut which would put it into “recording mode” and the user can type their new shortcut and it will be visible where the default shortcut keys were written and after typing the shortcut, the user can click on that shortcut row again to exit out of shortcut mode.

**I will write in more detail if I get a "good-to-go" from a maintainer.**

### Error handling
If a shortcut is an invalid shortcut then an error will be shown or if the shortcut is already being used then an error would be shown.

### Possible problems I'm thinking I would face

- Since Excalidraw is using i18n, and I cannot add a translation directly. I wanted to know how I should go about doing that.
- I saw a [command-palette branch](https://github.com/excalidraw/excalidraw/tree/command-palette) made by an Excalidraw member which is just going to be merged and I think it might affect and cause some conflicts in my work so I wanted some advice on whether to wait for that branch to be merged or start working immediately.


Any kind of help is always appreciated!

**Comment by viraptor** [comment]
FWIW, the command-palette branch seems to be gone now (maybe merged?)
The idea here sounds good to me @louremipsum An alternative could be setting the keybindings over the icon itself. (right-click -> set keybinding -> record the key press) It could be easier than hunting each action down a long list of actions without context. I don't know if that's harder to implement or not.

**Comment by louremipsum** [comment]
> @dwelle has the redesign mentioned in https://github.com/excalidraw/excalidraw/issues/3852#issuecomment-884037152 already happened?
(possibly not and something like would have to get extracted to something more global)

I guess I would have to wait for a confirmation from @dwelle before doing anything since I don't know if the team is planning on doing it on their own with their plan of action. 
As of now, I think since it would require much refactoring, I think maybe this might be handled by the maintainers only? Let's see

**Comment by wesleyverbruggen** [comment]
It would be extremely convenient to have this added. I use a pen on my Surface Laptop Studio, which hides my keyboard during drawing. To be still able to use shortcuts I use an external keypad. However, the current fixed keybindings are not very practical for my use case. Being able to adjust the keybindings would be amazing.

@dwelle can you inform @louremipsum at this moment in time?

**Comment by dwelle** [comment]
we are indeed planning keyboard customization, but there's a series of steps we need to make, one of which is rewriting the keyboard handling from the ground up, which won't be super easy even if we decided to use an existing library to handle keyboard (which we might not be able to use due to how the keyboard handling will be designed).

We could hack something together as an interim solution, but then would need migration paths for the final one. It'd will not be simple as a lot of keybindings aren't made through Actions.

**Comment by louremipsum** [comment]
@dwelle (Sorry for tagging you again) I'm actually free right now and would be more than happy to help you in the long term for building this solution. 
Although it would be much understandable if the team wants to handle this internally in a later period of time. (if so then you can add an appropriate label for this issue for people who would stumble upon this issue in the future)

**Comment by waLLxAck** [comment]
Has there been any progress here? An interim solution was discussed back in May 30, 2024. Soon it will be a year since...

**Comment by hpvd** [comment]
is there any news on changing keyboard short cuts within excalidraw?
This is IMHO a important one, since it effects every user and would not only enhance_
- comfort but also
- productivity and
- accessibility of
excalidraw.

**Comment by ryan-cm** [comment]
Following up on this. Would also love this feature.

**Comment by yuriever** [comment]
+1

**Comment by AhmedHawater2003** [comment]
+1

**Comment by Sebert81** [comment]
+ 1
It is very annoying that you cant customize the keyboard shortcuts:
e.g. If you use a QWERTZ keyboard the > & < keys are mapped to one key, so you can only increase the font size with the CTRL+SHIFT shortcut, but never decrease.

**Comment by viraptor** [comment]
To add more context, what I really want is to create key bindings for a few things that are either in unconvenient places, or just not available right now. I'd like to clear a few things that are in the way, but I never use, and instead have a tight cluster of box/arrow/colours/text keyboard actions for fast diagraming.

**Comment by gioele-prete-tecnosophical** [comment]
Is this still going to be implemented?
